### ISSUE TITLE
Take advantage of algorihmic differences between first-from-ordered and ordered.

### ISSUE THREAD

**Issue by JonHanna** [dotnet/corefx#2400]
The fact of something being ordered can have implications for subsequent operations.

The ordering Linq methods (`OrderBy`, `ThenBy`, etc.) operate first in O(1) time and space to make a promise of a subsequent O(n log n) sort in O(n) space. In many case though the following operation does not need all of the promised operation.

`.OrderBy(λ).First()`, `.OrderBy(λ).First(λ)`, and so on request that a sequence be sorted (O(n log n) time/O(n) space) and then the first element taken (O(1) time and space) which is in total an O(n log n) time/O(n) space operation. However, it is reducible to the O(n) time/O(1) space operation of identifying just the item in question which can be done in O(n) time and O(1) space.

As `First`, `Last` and their variants are particularly common with ordered sequences (in some cases not having any meaning without an order), I think this would be a worth-while optimisation.

There are other optimisations available, such as all aggregate operations (`Sum` etc.) could be moved from O(n log n) time to O(n) time simply by extracting the source and ignoring the ordering, this may not be as common a case as to make it worthwhile. One could argue that the user should have not put the ordering in there in the first place, and while the cost of checking for an ordered source is slightly, it's not worth paying it on all aggregate operations to benefit a flawed case.

`Reverse` is somewhere in the middle. It's reasonable to expect someone might construct an ordered sequence and then `Reverse` in some cases. It's got subtleties (one cannot simply reverse all `OrderBy` to `OrderByDescending` and vice-versa because `OrderBy` is documented as performing stable ordering in linq-2-objects and therefore reversed must have reverse source ordering on equivalent items) but wouldn't be hard, but it's less clear whether the pay off would be worth it.

In the more useful cases of `First` etc. there is a fly in the ointment in that this would affect the number of times both key selectors for ordering and predicates for `First` etc. were called. As per #2349 this means side effects on `Func`s would be affected, though perhaps in a way that was less predictable to the user than with that issue anyway.

As such, changing the behaviour of `First` in this case would require either a decision in favour of the change that issue would allow for, or a decision that this case was less harmed by such changing side-effects.

In summary. I propose that `OrderBy(…).First(…)` and its cousins be optimised, and the trade-offs of optimising `OrderBy(…).Reverse()` be considered.

**Comment by stephentoub** [comment]
Thanks for starting the discussion.

On one hand, I always gravitate towards things that make code faster. On the other hand, it feels like a very slippery slope towards a full query plan analyzer, and for cases that would hopefully be simply written differently by the developer in the first place, e.g. using Min(...) or Max(...) instead of OrderBy(...).First.  If we're missing an overload of Max/Min that would make this more feasible (e.g. a MaxBy as outlined in https://github.com/dotnet/corefx/issues/2146), then I'd prefer we invest in such generally useful overloads rather than adding more and more code to the codebase to workaround misemployed patterns like this.

**Comment by stephentoub** [comment]
cc: @VSadov

**Comment by VSadov** [comment]
Yes. There is a concern that this is a solution for a missing Min/Max overload. And there is a slippery slope concern.
However this particular case, the change seems not very big (especially the "analysis" part) and the combination of OrderBy+First could come as a result of query composition. Also OrderBy has direct language/IDE support in C#/VB query comprehensions. The new overloads for Min/Max would not have that.
I tend to think this change in principle is actually simpler than adding a new operator or an overload when everything is considered. 
I have not had a chance to look into every detail of the implementation though.

**Comment by hackcraft** [comment]
> On the other hand, it feels like a very slippery slope towards a full query plan analyzer

That's sort of what I was thinking of in mentioning, but dismissing, optimising for aggregate methods. All of them could be made faster for ordered sequences by just adding:

    var ordered = source as OrderedEnumerable<TSource>;
    if (ordered != null) return source = ordered.source;

But that would definitely be a case of, as you say "cases that would hopefully be simply written differently by the developer in the first place". These sort of tests for opportunities to optimise for a particular case have their costs, and while that case is only spending nanoseconds for what can be a big win, why should the person who made the wise choice spend even a picosecond to benefit the person who should have done things differently?

However, I think there is a natural sequence to `OrderBy(…).First()`. If we ignore cases where we expect there to be only one matching element anyway (in which case `First` is really an optimised `Single`) or where we don't care which matching element we get, then `First` is given its meaning by the previous `OrderBy`.

Likewise, the same is true for `OrderBy(…).Skip(…)`, `OrderBy(…).Take(…)` and in particular `OrderBy(…).Skip(…).Take(…)` as the classic paging pattern, and there are opportunities to reduce the time from O(n log n) to O(n + k log k) for the k taken, so I think I'll definitely look at that before taking off the WIP tag.

> If we're missing an overload of Max/Min that would make this more feasible (e.g. a MaxBy as outlined in #2146), then I'd prefer we invest in such generally useful overloads

It was in offering a proposed implementation to that, that reminded me of this case. However while `source.OrderBy(selector).First() == source.MinBy(selector)` there isn't as clear an equivalence for `source.OrderBy(selector0).ThenBy(selector1).ThenByDescending(selector2)`.

(A flip side to this though is that optimising `OrderBy(…).First()` would reduce the need for `MinBy`).

>  than adding more and more code to the codebase to workaround misemployed patterns like this

I don't think these are misemployed patterns, for the reasons I've discussed above. It's that I think `.OrderBy(…).Sum()` is misemployed that I don't think it would pay to optimise for. It's that I'm not sure whether `OrderBy(…).Reverse` is misemployed or not (because it's subtly different to `OrderByDescending(…)`) that I'm not sure whether it would be worth optimising.

**Comment by hackcraft** [comment]
@VSadov 
> However this particular case, the change seems not very big

What do you think of the fact that it changes side-effects? I'm inclined to think it's not a problem myself, but it is something that makes the change not a "pure" optimisation.

**Comment by VSadov** [comment]
For the sideeffects in queries. And this is just my opinion at the moment  :-)  :

I think the order of execution of iterating operations and user code is important - i.e. if something runs as a part of GetEnumerator/MoveNext/Current/Dispose, it should continue happening there since the distinction is somewhat defined or predictable, directly or indirectly. If something throws and user has some compensating code, moving sideeffects to a different stage of iterating is likely to be breaking.

On the other hand I am much less concerned with how many times we call comparers or GetHashcode or how far we dig into source sequence. It seems much more likely that these operations should be tolerant to (but not dependent on) redundancy in the implementation of query operators.
From this prospective, changes like short-circuiting would seem to be ok.

I have not looked at this change deep enough to see what kind of sideeffects change we can expect from this optimization.

**Comment by VSadov** [comment]
As a theoretical example - Another possible optimization around ordering could be, since we are using quick sort, is deferring the sorting of the other half until we MoveNext into it. What if we need only the first n elements? We do not need to sort the rest.
The optimization is observable through how many times we call comparers, but I think dependency on that is unreasonable and unlikely, and therefore partial sort deferring could be a valid optimization, if can be done cheap enough.

**Comment by stephentoub** [comment]
> However while source.OrderBy(selector).First() == source.MinBy(selector) there isn't as clear an equivalence for source.OrderBy(selector0).ThenBy(selector1).ThenByDescending(selector2).First

Hmm, that's a good point.  I can buy that, as well as the OrderBy(...).Take(...) example.  If supporting OrderBy(...).First naturally came from supporting such things, seems fairly reasonable then.

**Comment by hackcraft** [comment]
> if something runs as a part of GetEnumerator/MoveNext/Current/Dispose, it should continue happening there since the distinction is somewhat defined or predictable, directly or indirectly

I agree. I proposed a performance improvement to `Skip` previously that broke this sequence in the case of exceptions, and agreed that it was therefore inappropriate. (That case is fixable, but I haven't yet checked if the cost of ensuring it doesn't break that pattern outweighs the improvements. If it's still an improvement I'll be proposing it along with measures of the improvements made).

> It seems much more likely that these operations should be tolerant to (but not dependent on) redundancy in the implementation of query operators.
From this prospective, changes like short-circuiting would seem to be ok.

See also: #2349 

> Another possible optimization around ordering could be, since we are using quick sort, deferring the sorting of the other half until we MoveNext into it. What if we need only the first n elements? We do not need to sort the rest.

I wonder how much it would complicate the implementation to `yield` a sorted partition once sorted (we'd obviously want to do the lower partition before the higher), and the impact of that on the case of all elements being taken. I'd guess it wouldn't be worth it, but the only way to know is to try and then measure.

However, in the case of `Take` and `Skip` we have been given information that we definitely only want a certain range, so I think that's definitely worth looking at, and I shall do so soon.

**Comment by hackcraft** [comment]
> I have not looked at this change deep enough to see what kind of sideeffects change we can expect from this optimization.

Oh. Take a look at the test cases in the proposed implementations, they document the side-effect changes between those marked `[ActiveIssue("Demonstrates old behaviour.")]` and those that are not. (`Last` is unchanged in this regard).

I would note that it's important that ordering key selectors are only ever called once per element, as sometimes people use non-deterministic selectors such as `OrderBy(i => rnd.Next()` for a quick-and-dirty shuffle (flawed in a few ways, suitable only for throwaway scratch code, but certainly done) which could be nasty if key selectors aren't hit. The change in the PR skips calling the selector in some cases, but never hits it more than once for a given item.

**Comment by Clockwork-Muse** [comment]
Side note: PLINQ already drops ordering for aggregate operations.

**Comment by hackcraft** [comment]
@Clockwork-Muse that's interesting, though I don't think it's a complete like-with-like correspondence:

Plain linq-to-objects generally preserves order from further up the path. Aggregate methods of course destroy that order because they produce a single result, but in cases which could have an order, maintaining that order "just happens".

So if an ordering is done that isn't needed, the waste is only that inherent to the ordering itself.

Parallel operations generally destroy order. If you want to maintain an order, then the parallel operations have to take on the extra burden of doing so.

So if an ordering is done that isn't needed, the waste is not just that of the ordering itself, but of impacting upon subsequent operations not being able to take the most efficient path.

So the benefit in the parallel case of dropping unnecessary orderings is greater than in the linear case.

Not that there's no benefit for linear cases, or that it would be hard to do (code for that is a couple of comments back), or conversely not that with the parallel case we couldn't argue that the calling code just shouldn't have done that ordering, but the two cases aren't strictly analogous.

The example does though show that looking at the effects of ordering on an operation and adjusting accordingly has been considered a good policy in the past.

Whether dropping ordering in linq2objects aggregates is considered wise or not, I think the gains are greater, and the costs more justifiable, where ordering affects an outcome in a way that is more reasonable for the calling code to make use of (`First`, `Take`, etc).

On which basis, the PR now covers `Take`, `Skip`, `ElementAt` and `ElementAtOrDefault`. All remain O(n) in space, but `Take` and `Skip` are O(n + k log k) in time where `k` elements are the number retrieved (the added n there is part of the O(n log n) of the original, so there's no case where the O(n log n) comes out as cheaper than the O(n + k log k) of the new approach) and the `ElementAt` cases are O(n) in time in the best case (a rare worse-case is O(n²) but the previous approach would have been O(n²) in that case too).

**Comment by hackcraft** [comment]
@VSadov 
> Another possible optimization around ordering could be, since we are using quick sort, deferring the sorting of the other half until we MoveNext into it. What if we need only the first n elements? We do not need to sort the rest.

There's a lot of further possible optimisations around this, and the reason the PR stayed marked [WIP] for so long was that I was examining too many of them, but I've reigned it in for now.

A problem with this suggestion is that since we can't enumerate until we have results, we have to either always go for the left-most partition first, or else have rather complicated tracking of which results have been sent already. If we always go for the left-most partition we'd have to undo the way that the current implementation only recurses on the smaller partition; essentially Sedgewick's improvement of tail-calling on the larger partition, but with the tail-call elimination done by manual conversion to iterative. Without this almost all cases would be a bit slower, worse cases would be much worse, and the stack-depth bounded by O(n) rather than O(log n).

It could though work well if we stored the details of still-to-sort partitions in a stack of left-right pairs within the method, and made the entire thing iterative. That in itself is a valid optimisation in its own right, and would counter much of the down-side of yielding as we go.

Yielding as we go would still likely be a bit slower than preparing the entire buffer first in cases where we end up wanting the entire buffer, though that's a bit hard to predict (since the current path does still end up having to iterate through that buffer to pass it out after all).

In all, I think this is worth looking at, and indeed I'm going to look at it, but I'm going to do it separate to the PR I submitted for this issue.

**Comment by JonHanna** [comment]
@VSadov

> Another possible optimization around ordering could be, since we are using quick sort, deferring the sorting of the other half until we MoveNext into it. What if we need only the first n elements? We do not need to sort the rest.

I did experiment with this a while ago, and forgot to report back. In all it didn't seem to be worth it. Forcing the "left" side to always be sorted first can result in much worse behaviour in some cases, and any attempts to be clever about keeping track of where was sorted so that it could start yielding once it had a batch of sorted values that included the "current" point resulted in too much complication that slowed down any case that didn't explicitly benefit from the optimisation. Either way it was a loss to any case that did use all elements.
